secure/storage/vault/src/lib.rs

```diff
@@ -391,6 +406,134 @@ impl Client {
+///       "min_decryption_version":1,
+///       "min_encryption_version":0,
+///       "latest_version":1,
+///       "archive_version":1,
+///       "archive_min_version":0,
+///       "min_available_version":0,
+///       "deletion_allowed":false,
+///       "convergent_version":0,
+///       "type":2,
+///       "backup_info":{
+///          "time":"2020-05-29T06:28:48.2937047Z",
+///          "version":1
+///       },
+///       "restore_info":null,
+///       "allow_plaintext_backup":true,
+///       "version_template":"",
+///       "storage_prefix":""
+///    }
+/// }
+///
+/// This is intended to be a very simple application of it only for the purpose of introducing a
+/// single key with no versioning history into Vault. This is /only/ for test purposes and not
+/// intended for production use cases.
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
+pub struct KeyBackup {
+    policy: KeyBackupPolicy,
+}
+
+impl KeyBackup {
+    pub fn new(key: &Ed25519PrivateKey) -> Self {
+        let pub_key = key.public_key();
+        let mut key_bytes = key.to_bytes().to_vec();
+        let pub_key_bytes = pub_key.to_bytes().to_vec();
+        key_bytes.extend(&pub_key_bytes);
+
+        let now = chrono::Utc::now();
+        let time_as_str = now.to_rfc3339();
+
+        let mut info = KeyBackupInfo::default();
+        info.key = Some(base64::encode(key_bytes));
+        info.public_key = Some(base64::encode(pub_key_bytes));
+        info.creation_time = now.timestamp_subsec_millis();
+        info.time = time_as_str.clone();
+
+        let mut key_backup = Self {
+            policy: KeyBackupPolicy::default(),
+        };
+
+        key_backup.policy.exportable = true;
+        key_backup.policy.min_decryption_version = 1;
+        key_backup.policy.latest_version = 1;
+        key_backup.policy.archive_version = 1;
+        key_backup.policy.backup_type = 2;
+        key_backup.policy.keys.insert(1, info);
+        key_backup.policy.backup_info.time = time_as_str;
+        key_backup.policy.backup_info.version = 1;
+        key_backup
+    }
+}
+
+#[derive(Debug, Default, Deserialize, PartialEq, Serialize)]
+pub struct KeyBackupPolicy {
+    name: String,
+    keys: BTreeMap<u32, KeyBackupInfo>,
+    derived: bool,
+    kdf: u32,
+    convergent_encryption: bool,
+    exportable: bool,
+    min_decryption_version: u32,
+    min_encryption_version: u32,
+    latest_version: u32,
+    archive_version: u32,
+    archive_min_version: u32,
+    min_available_version: u32,
+    deletion_allowed: bool,
+    convergent_version: u32,
+    #[serde(rename = "type")]
+    backup_type: u32,
+    backup_info: BackupInfo,
+    restore_info: Option<()>,
+    allow_plaintext_backup: bool,
+    version_template: String,
+    storage_prefix: String,
+}
+
+#[derive(Debug, Default, Deserialize, PartialEq, Serialize)]
+pub struct KeyBackupInfo {
+    key: Option<String>,
+    hhmac_key: Option<String>,
+    time: String,
+    ec_x: Option<String>,
```

Comment:
I'm wondering if it's strictly necessary to define all of these fields when sending the "restore" request? I know "backup" returns them as null, so it probably makes sense to define them, but there's a lot of unused ones here.. especially as we aren't calling "backup" ourselves. Either way, this is probably okay!

Reply:
At some point, I just left well enough alone. I already trimmed a lot of content from this :)